[BE] refactor: 도메인 코드 위주의 코드 리팩토링 #354

Byesol opened the pull request:
<!-- PR 제목 예시
ex. [BE] 로그인 API 구현
ex. [AN] 페이지 구현 -->

## 📌 변경 내용 & 이유

1. Hearit 파일 URL 관리 방식 개선
기존에 Hearit 엔티티에 흩어져 있던 originalAudioUrl, shortAudioUrl, scriptUrl 필드를 FileUrls라는 값 객체(Value Object)로 통합했습니다. 파일 경로와 관련된 유효성 검사 로직을 FileNameValidator에서 FileUrls와 FileType 도메인 객체 내부로 이동시켜 도메인의 응집도를 높임. 그에 따른 전체 테스트 코드들도 수정됨.
 단, 서비스 부분은 사나 pr 과의 충돌을 피하기 위해 최대한 수정을 하지 않았습니다.

2. RecommendHearit의 Hearit 참조 방식 변경
//TODO로 써져 있던 기존 Long hearitId로 관리하던 간접 참조를 Hearit hearit 객체를 직접 참조하는 방식으로 변경

3. 탐색 점수 계산기 인터페이스 분리
ExploreScoreCalculator를 클래스에서 인터페이스로 분리하고, DefaultExploreScoreCalculator(이름 추천 바람) 구현체를 만듬

4. API 요청 파라미터 객체화
 커서 기반 페이징 파라미터(cursorId, size)를 CursorRequest라는 DTO로 묶어서 관리

5. N+1 성능 문제 해결
 지연/즉시 로딩 시 발생할 수 있는 N+1 문제를 해결하기 위해 RecommendHearitRepository, HearitKeywordRepository 등의 여러 쿼리에 FETCH JOIN을 적용

6. 도메인 코드 내에 eqauls, hashCode 추가


## 📸 스크린샷 (선택)
> UI 변경이 있을 경우 첨부

## 🧪 테스트 방법 (선택)
<!-- 코드 리뷰하는 팀원이 돌려볼 수 있도록 작성해주기 
ex. ReservationTest의 XXXX 테스트메서드로 확인 
ex. 어떤화면의 어떤기능에서 확인 -->

## 📢 논의하고 싶은 내용
<!-- 이러이러해서 이부분을 이러이러하게 구현했는데 이방식이 괜찮은지? -->
기존 Hearit 코드 내의 FileUrl vo를 추가하는 과정에서 이제

```
    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "originalAudioUrl", column = @Column(name = "original_audio_url")),
            @AttributeOverride(name = "shortAudioUrl", column = @Column(name = "short_audio_url")),
            @AttributeOverride(name = "scriptUrl", column = @Column(name = "script_url"))
    })
    private FileUrls fileUrls;
```
위와 같이 코드를 작성했는데 그 이유는 이제 기존 vo로 묶으며  기존 데이터의 column 명과 달라져 모든 데이터를 피하고, 또한, 굳이 column명에 vo명이 안 들어가도 된다고 생각했습니다.
## 🧩 관련 이슈
<!-- 이슈 태그: #123 -->
<!-- 이슈 닫기: close #123 -->
#354 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- 신기능
  - 미디어 파일 URL을 통합한 값 객체로 관리하여 조회·업데이트가 더 명확해졌습니다.
- 성능 개선
  - 추천·키워드 조회에서 연관 데이터 일괄 로딩을 도입해 N+1 쿼리를 줄였습니다.
- 리팩터
  - 추천 제공 방식을 전략 기반으로 단순화하고 식별자 대신 객체 연관을 직접 사용하도록 전환했습니다.
- 안정성 향상
  - 여러 도메인에 입력 검증 및 equals/hashCode 구현을 추가해 null 처리와 동등성 판단을 강화했습니다.
- 테스트
  - 테스트 데이터와 검증을 실제 경로 기반 값으로 변경하고 관련 테스트를 정리했습니다.
- 제거
  - 무작위 조회 기능(랜덤 선택)이 제거되었습니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Byesol committed:
[BE] refactor: hearit Url 들을 VO로 변경 ##354

Byesol committed:
[BE] refactor: FileUrls로 validate 책임 변경  #354

Byesol committed:
[BE] refactor: fileType 으로 update와 get 책임변경 #354

Byesol committed:
[BE] refactor: accessLevel 설정 #354

Byesol committed:
[BE] refactor: source이 column 명, nullable 설정 #354

Byesol committed:
[BE] refactor: fileNameValidator 복구 #354

Byesol committed:
[BE] test: fileName 규칙 추가에 따른 테스트 데이터 변경 #354

Byesol committed:
[BE] refactor: 파일 경로 유효성 검사 로직 도메인 내로 이동함에 따른 테스트 코드 변경 #354

Byesol committed:
[BE] refactor: 추천 히어릿 간접 참조를 직접 참조로 변경 #354

Byesol committed:
[BE] refactor: calculator를 인터페이스로 분리 #354

Byesol committed:
[BE] refactor: 컨트롤러 요청 시 cursorRequest 적용 #354

Byesol committed:
[BE] refactor: 사용하지 않는 메서드 제거 #354

Byesol committed:
[BE] refactor: fetch join 설정  #354

Byesol committed:
[BE] refactor: 도메인 코드 내에 eqauls, hashCode 추가  #354

Byesol committed:
[BE] refactor: url 의 nullable 을 false 처리 #354

JO-eusan commented on backend/src/main/java/com/onair/hearit/admin/application/AdminHearitService.java:
```diff
@@ -73,10 +72,10 @@ private Map<Long, List<KeywordInHearit>> mapKeywordsByHearitId(List<HearitKeywor
-                request.shortAudio().getOriginalFilename(),
-                request.scriptFile().getOriginalFilename());
+//        FileNameValidator.validateAll(
+//                request.originalAudio().getOriginalFilename(),
```
`AdminHearitService` 클래스 함수는 제 pr에서 봐주시면 됩니다.

(https://github.com/woowacourse-teams/2025-hEARit/blob/bb96ac83befdafc154170e113a4e301b08ac4508/backend/src/main/java/com/onair/hearit/admin/application/AdminHearitService.java)

JO-eusan commented on backend/src/main/java/com/onair/hearit/admin/application/AdminRecommendHearitService.java:
```diff
@@ -63,7 +63,7 @@ public List<MonthlyRecommendedHearitResponse> getMonthRecommendedHearit(Integer
             List<Long> hearitIds = monthlyRecommendHearits.stream()
                     .filter(recommendHearit -> recommendHearit.getRecommendDate().isEqual(target))
-                    .map(RecommendHearit::getHearitId)
+                    .map(recommendHearit -> recommendHearit.getHearit().getId())
```
이 과정이 결론적으로 해당 히어릿 Id를 뽑아서 `List<Hearit>`을 만드는 거니까, 이제는 추천 히어릿 Id를 받아오는 과정을 제거해서 바로 repository에서 히어릿 자체를 조회할 수 있을 것 같습니다!

이 부분은 제 pr과 충돌이 불가피할 것 같아서 merge 같이 하면 될 것 같습니다.

Byesol replied:
일단 놔둘께요!

JO-eusan commented on backend/src/main/java/com/onair/hearit/domain/FileType.java:
```diff
@@ -10,50 +11,50 @@ public enum FileType {
-        if (!filename.startsWith(prefix)) {
-            throw new InvalidInputException(this.name() + "의 파일명은 '" + prefix + "'로 시작해야 합니다.");
+    public void validUrlName(String filename) {
+        System.out.println("filename = " + filename);
```
디버깅을 위한 `println`은 제거하면 좋을 것 같습니다!

Byesol replied:
제거했습니다!

JO-eusan commented on backend/src/main/java/com/onair/hearit/application/explore/DefaultExploreScoreCalculator.java:
```diff
@@ -0,0 +1,26 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class DefaultExploreScoreCalculator implements ExploreScoreCalculator {
```
인터페이스 이름에 `Calculator`보다는 `Strategy`가 어떨까요? 사실 계산 방식이 바뀌기 보다는 완전히 로직이 변경될 것 같아서요.

- 인터페이스: ExploreScoreStrategy
- 구현체: ExploreFactorScoring

yuyoungrhee replied:
저는 Calculator는 괜찮다고생각하는데 계산 방식이 바뀌는 거 아닌가요? 
결국 점수 기반이면 Calculator가 맞는 거 같고, 다른전략이면 아예 ExploreStrategy로 더 크게 추상화가 돼야할 거 같은 느낌.
지금 Score를유지할거면 Calculator가 맞다고 생각함요

Byesol replied:
해당 부분은 대면 논의 후 일단, calulator 인터페이스 분리가 아닌 상위클래스를 두는 게 맞다고 결론이 내려져 인터페이스를 제거하고 다른 이슈를 통해 상위클래스 도입하기로 했습니다.

JO-eusan commented on backend/src/main/java/com/onair/hearit/application/explore/HearitExploreService.java:
```diff
@@ -37,17 +38,17 @@ public class HearitExploreService {
 
-    public CursorResponse<ExploredHearitResponse> getExploredHearits(Long memberId, Long cursorId, int size) {
-        if (isFirstExploreRequest(cursorId)) {
+    public CursorResponse<ExploredHearitResponse> getExploredHearits(Long memberId, CursorRequest request) {
```
일관되게 변경 요청/응답 변경 good 👍🏻

JO-eusan commented on backend/src/main/java/com/onair/hearit/application/DbRecommendHearitProvider.java:
```diff
@@ -20,9 +20,9 @@ public class DbRecommendHearitProvider implements RecommendHearitProvider {
-        List<Long> recommendHearitIds = recommendHearits.stream()
-                .map(RecommendHearit::getHearitId).toList();
-        return hearitRepository.findAllByIdInWithCategory(recommendHearitIds);
+        return recommendHearits.stream()
```
추천 히어릿을 제공하는 전략도 이름을 다음과 같은 방식으로 변경하는 거 어떤가요?

- 인터페이스: RecommendHearitStrategy
- 구현체: RecommendStoredHearitProviding

yuyoungrhee replied:
이건 좋음요 근데 구현체 그냥 FixedRecommendHearitStrategy 어떰

Byesol replied:
구현체 - FixedRecommendHearitStrategy
인터페이스 - RecommendHearitStrategy 로 수정했습니다!

JO-eusan commented on backend/src/main/java/com/onair/hearit/domain/Bookmark.java:
```diff
@@ -47,4 +48,23 @@ public Bookmark(Member member, Hearit hearit) {
     }
+
+    @Override
+    public boolean equals(Object o) {
```
롬복에 `@EqualsAndHashCode` 어노테이션이 있다는데 이를 적용해보면 어떨까요?
팀에서 롬복을 사용하는 이유가 프로덕션 코드에 집중하기 위한 거니까 이 목적이랑 잘 맞을 것 같아요.

Byesol replied:
팀 내 논의 결과, equals hashcode의 경우 프로덕션 내에서 유지하기로 합의되었습니다!

JO-eusan commented on backend/src/main/java/com/onair/hearit/domain/FileUrls.java:
```diff
@@ -0,0 +1,83 @@
+        this.scriptUrl = scriptUrl;
+    }
+
+    private void validateAll(String originalAudioPath, String shortAudioPath, String scriptFilePath) {
```
`validateAll` 보다는 `validateFilePath`가 더 명확할 것 같습니다.

yuyoungrhee replied:
하나의 파일Path만 검증하는 듯한 느낌이 들어서 validateAllFilePath 어떠삼요?

Byesol replied:
validateAllFilePath로 수정했습니다!

JO-eusan commented on backend/src/main/java/com/onair/hearit/domain/Hearit.java:
```diff
@@ -41,14 +45,16 @@ public class Hearit {
-    @Column(name = "script_url", nullable = false)
-    private String scriptUrl;
+    @Embedded
+    @AttributeOverrides({
```
@AttributeOverrides를 사용하신 이유가 있나요? 어떤 어노테이션일까요?
Embeddable 타입으로 VO 객체 내에 이렇게 작성하면 안되는건지 궁금합니다.

```java
@Column(name = "original_audio_url", nullable = false)
    private String originalAudioUrl;

    @Column(name = "short_audio_url", nullable = false)
    private String shortAudioUrl;

    @Column(name = "script_url", nullable = false)
    private String scriptUrl;
```

Byesol replied:
아 embedeabble 타입이라도 해당 방식처럼 작성하는 게 훨씬 낫겠네요. 수정할께요!

JO-eusan commented on backend/src/main/java/com/onair/hearit/domain/Hearit.java:
```diff
@@ -71,9 +77,8 @@ public Hearit(String title, String summary, Integer playTime, String originalAud
-        this.originalAudioUrl = originalAudioUrl;
-        this.shortAudioUrl = shortAudioUrl;
-        this.scriptUrl = scriptUrl;
+        System.out.println("originalAudioUrl = " + originalAudioUrl);
```
`println` 제거해주세요

Byesol replied:
제거했습니다

JO-eusan commented on backend/src/main/java/com/onair/hearit/domain/RecommendHearit.java:
```diff
@@ -23,25 +25,44 @@ public class RecommendHearit {
-    private void validateFields(Long hearitId, LocalDate recommendDate) {
-        if (hearitId == null) {
-            throw new InvalidInputException("hearit Id는 null일 수 없습니다.");
+    private void validateFields(Hearit hearit, LocalDate recommendDate) {
```
validateNull은 어떨까요?

Byesol replied:
반영했어요!

JO-eusan commented on backend/src/main/java/com/onair/hearit/infrastructure/HearitKeywordRepository.java:
```diff
@@ -12,7 +12,8 @@ public interface HearitKeywordRepository extends JpaRepository<HearitKeyword, Lo
                 SELECT hk
                 FROM HearitKeyword hk
-                JOIN hk.keyword k
+               JOIN FETCH hk.hearit h
```
미세하게 띄어쓰기가 다른 행들과 맞지 않는것 같습니다.

Byesol replied:
감사해용

JO-eusan commented on backend/src/main/java/com/onair/hearit/infrastructure/RecommendHearitRepository.java:
```diff
@@ -11,24 +11,26 @@
+    Optional<RecommendHearit> findRecentByHearitId(@Param("hearitId") Long hearitId);
 
-    List<RecommendHearit> findByRecommendDateIsBetween(LocalDate recommendDateAfter, LocalDate recommendDateBefore);
+    @Query("SELECT rh FROM RecommendHearit rh JOIN FETCH rh.hearit WHERE rh.recommendDate BETWEEN :from AND :to")
```
이 부분 쿼리 줄바꿈이 필요할 것 같습니다.

Byesol replied:
했습니당

JO-eusan reviewed (changes requested):
벡터, 리팩터링 범위가 넓었는데 꼼꼼하게 잘 해주셨네용!
몇 가지 커멘트 남겨두었으니 확인 부탁드립니다.

Byesol committed:
[BE] refactor: 디버깅용 sout 및 주석제거  #354

Byesol committed:
[BE] refactor: ManytoOne에 lazy Loading 적용 #354

Byesol commented on backend/src/main/java/com/onair/hearit/domain/RecommendHearit.java:
```diff
@@ -23,25 +25,44 @@ public class RecommendHearit {
     @JoinColumn(name = "hearit_id", nullable = false)
-    private Long hearitId;
+    private Hearit hearit;
 
```
lazy loading 적용했어요!

Byesol committed:
[BE] refactor: 메서드명 의미에 맞게 수정 #354

Byesol committed:
[BE] refactor: 값 객체 내부에서 column 명을 정의하도록 수정 #354

Byesol committed:
[BE] refactor: 쿼리 정렬 #354

Byesol committed:
[BE] refactor: 북마크의 멤버 lazy loading 제거 #354

Byesol committed:
[BE] refactor: 클래스명 목적에 맞게 변경 #354

Byesol committed:
[BE] refactor: 필요없는 인터페이스 제거 #354

Byesol committed:
[BE] refactor: 필요없는 인터페이스 제거 #354

gabean13 commented on backend/src/main/java/com/onair/hearit/application/FixedRecommendHearitStrategy.java:
```diff
@@ -3,7 +3,7 @@
 import java.util.List;
 
-public interface RecommendHearitProvider {
+public interface FixedRecommendHearitStrategy {
```
얘는 그대로 RecommendHearitStrategy로 두는게 낫지 않나용?
Fixed를 붙힌 이유가 궁금합니다.

Byesol replied:
다시 보니 저도 이유를 모르겠네요.. ㅠ
구현체 - FixedRecommendHearitStrategy
인터페이스 - RecommendHearitStrategy 로 수정했습니다!

gabean13 commented on backend/src/main/java/com/onair/hearit/application/FixedRecommendedHearitProviding.java:
```diff
@@ -12,17 +12,17 @@
 @Service
 @RequiredArgsConstructor
-public class DbRecommendHearitProvider implements RecommendHearitProvider {
+public class FixedRecommendedHearitProviding implements FixedRecommendHearitStrategy {
```
얘는 왜 Providing인가염?? 차라리 얘를 FixedRecommendHearitStrategy로 두는거 어때요 interface를 RecommendHearitStrategy로 두고

Byesol replied:
구현체 - FixedRecommendHearitStrategy
인터페이스 - RecommendHearitStrategy 로 수정했습니다!

gabean13 commented on backend/src/main/java/com/onair/hearit/domain/FileUrls.java:
```diff
@@ -0,0 +1,89 @@
+        this.originalAudioUrl = originalAudioUrl;
+        this.shortAudioUrl = shortAudioUrl;
+        this.scriptUrl = scriptUrl;
+    }
```
왜 여기에는 final이 있나욤..?

Byesol replied:
헉... 감사합니다 ㅠㅠ

gabean13 commented on backend/src/main/java/com/onair/hearit/dto/request/CursorRequest.java:
```diff
@@ -0,0 +1,17 @@
+) {
+    public CursorRequest {
+        if (cursorId < 0) {
+            throw new InvalidInputException("page는 0 이상이어야합니다.");
```
cursor 라고 해야함요

Byesol replied:
반영했습니다!

gabean13 commented on backend/src/main/java/com/onair/hearit/infrastructure/RecommendHearitRepository.java:
```diff
@@ -11,24 +11,30 @@
+                WHERE rh.recommendDate <= :recommendDate
+                ORDER BY rh.recommendDate DESC
+                LIMIT :size
+            """)
```
```suggestion
                SELECT rh
                FROM RecommendHearit rh
                JOIN FETCH rh.hearit
                JOIN FETCH rh.hearit.category
                WHERE rh.recommendDate <= :recommendDate
                ORDER BY rh.recommendDate DESC
                LIMIT :size
            """)
```

일케해야지 N+1 문제 해결할 수 있을 듯염?? 왜냐면 지금은 hearit까지만 땡겨오고 category는 나중에 getCateogry할때 select 보내요!!

쿼리문 비교

기존
```sql
Hibernate: 
    select
        rh1_0.id,
        rh1_0.hearit_id,
        h1_0.id,
        h1_0.category_id,
        h1_0.created_at,
        h1_0.original_audio_url,
        h1_0.script_url,
        h1_0.short_audio_url,
        h1_0.play_time,
        h1_0.summary,
        h1_0.title,
        rh1_0.recommend_date 
    from
        recommend_hearit rh1_0 
    join
        hearit h1_0 
            on h1_0.id=rh1_0.hearit_id 
    where
        rh1_0.recommend_date<=? 
    order by
        rh1_0.recommend_date desc 
    fetch
        first ? rows only
```
fetch join에 category를 추가했을 때
```sql
Hibernate: 
    select
        rh1_0.id,
        rh1_0.hearit_id,
        h1_0.id,
        h1_0.category_id,
        c1_0.id,
        c1_0.color_code,
        c1_0.name,
        h1_0.created_at,
        h1_0.original_audio_url,
        h1_0.script_url,
        h1_0.short_audio_url,
        h1_0.play_time,
        h1_0.summary,
        h1_0.title,
        rh1_0.recommend_date 
    from
        recommend_hearit rh1_0 
    join
        hearit h1_0 
            on h1_0.id=rh1_0.hearit_id 
    join
        category c1_0 
            on c1_0.id=h1_0.category_id 
    where
        rh1_0.recommend_date<=? 
    order by
        rh1_0.recommend_date desc 
    fetch
        first ? rows only
```

Byesol replied:
오. 이 repository는 아예 놓쳤었네요. 감사합니다! 수정했어요!

gabean13 commented on backend/src/main/java/com/onair/hearit/infrastructure/RecommendHearitRepository.java:
```diff
@@ -11,24 +11,30 @@
+    @Query("""
+                 SELECT rh FROM RecommendHearit rh
+                 JOIN FETCH rh.hearit
+                 WHERE rh.recommendDate BETWEEN :from AND :to
```
여기도 동일하게 Cateogory fetch join해야될것같아염

Byesol replied:
수정했습니다!

gabean13 reviewed (commented):
고생했어요 벡터

Byesol committed:
[BE] Merge branch 'develop-be' into refactor/#354-refactor-domain

# Conflicts:
#	backend/src/main/java/com/onair/hearit/admin/application/AdminHearitService.java
#	backend/src/main/java/com/onair/hearit/admin/application/AdminRecommendHearitService.java
#	backend/src/main/java/com/onair/hearit/admin/domain/FileNameValidator.java
#	backend/src/main/java/com/onair/hearit/application/HearitService.java
#	backend/src/main/java/com/onair/hearit/application/explore/HearitExploreService.java
#	backend/src/main/java/com/onair/hearit/dto/request/CursorRequest.java
#	backend/src/main/java/com/onair/hearit/presentation/HearitController.java
#	backend/src/test/java/com/onair/hearit/admin/presentation/AdminHearitControllerTest.java
#	backend/src/test/java/com/onair/hearit/application/HearitExploreServiceTest.java
#	backend/src/test/java/com/onair/hearit/application/HearitServiceTest.java

Byesol committed:
[BE] refactor: 병합으로 인한 코드 수정 #354

Byesol committed:
[BE] refactor: 도메인 내 null 검사 추가 #354

Byesol committed:
[BE] refactor: 필요없는 repository 변수 제거 #354

Byesol committed:
[BE] refactor: 메서드 파라미터의 final 제거 #354

Byesol committed:
[BE] refactor: fetch Join 적용 #354

Byesol committed:
[BE] refactor: hearit 직접 참조에 따른 리팩토링 #354

Byesol committed:
[BE] refactor: 클래스명 strategy로 변경 #354

Byesol committed:
[BE] refactor: 메서드 순서 변경 #354

Byesol reviewed (commented):
리뷰 반영완료했습니다!

Byesol committed:
[BE] refactor: db에 recommendHearit 일괄 저장하게 변경  #354

yuyoungrhee reviewed (approved):
리뷰 반영 확인했습니다! 수고하셨어요

gabean13 reviewed (approved):
고생했어요 벡터~

JO-eusan reviewed (approved):
벡터 연휴에 고생하셨어요! Approve 할게용
